### PR TITLE
 fix: add gap for future proposals and votes

### PR DESCRIPTION
**Pull request by KaoImin**
This PR does:
* add a gap for future proposals and votes
* fix a bug may cause commit error
* more stringent self-check logic

### CODE REVIEW

File: src/state/process.rs [hunk 1]
```diff
@@ -407,7 +408,11 @@ where
         // If the proposal epoch ID is higher than the current epoch ID or proposal epoch ID is
         // equal to the current epoch ID and the proposal round is higher than the current round,
         // cache it until that epoch ID.
-        if epoch_id > self.epoch_id || (epoch_id == self.epoch_id && round > self.round) {
+        if (epoch_id > self.epoch_id && self.epoch_id + FUTURE_EPOCH_GAP > epoch_id)
```
Comment (rev-chaos) [inline src/state/process.rs]: If epoch_id between (self.epoch_id, self.epoch_id + FUTURE_EPOCH_GAP)，but round is very large, it is also not allowed to insert into self.proposals. You should concern that. 
Besides, the proposals not satisfy your condition is not filtered.

File: src/state/process.rs [hunk 2]
```diff
@@ -702,21 +693,28 @@ where
...
         self.verify_address(&vote.voter, true)?;
-        self.votes
-            .insert_vote(signed_vote.get_hash(), signed_vote, vote.voter);
 
         // If the vote epoch ID is higher than the current epoch ID, cache it and rehandle it by
         // entering the epoch. Else if the vote epoch ID is equal to the current epoch ID
         // and the vote round is higher than the current round, cache it until that round
         // and precess it.
-        if epoch_id > self.epoch_id || (epoch_id == self.epoch_id && round > self.round) {
+        if (epoch_id > self.epoch_id && self.epoch_id + FUTURE_EPOCH_GAP > epoch_id)
```
Comment (rev-chaos) [inline src/state/process.rs]: Same problem.

File: src/state/process.rs (hunk 1 above)
Comment (KaoImin) [reply src/state/process.rs]: ok

File: src/state/process.rs [hunk 3]
```diff
@@ -26,6 +26,8 @@ use crate::{Codec, Consensus, ConsensusResult, Crypto, INIT_EPOCH_ID, INIT_ROUND
 
 const CHECK_EPOCH_SUCCESS: bool = true;
 const CHECK_EPOCH_FAILED: bool = false;
+const FUTURE_EPOCH_GAP: u64 = 5;
```
Comment (yejiayu) [inline src/state/process.rs]: Is this value defined in the Overlord protocol?

Comment (KaoImin) [reply src/state/process.rs]: yes

**Pull request merged**